searchable.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020-2022 CYBERCRYPT
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encryptonize
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/cyber-crypt-com/encryptonize-lib/crypto"
+)
+
+// Index contains a mapping from string label to SealedID.
+type Index struct {
+	mapping map[string]SealedID
+}
+
+// SealedID is an encrypted structure which defines an occurrence of a specific keyword in a specific ID.
+type SealedID struct {
+	ciphertext []byte
+	wrappedKey []byte
+}
+
+const MasterKeyLength = 32
+
+var ErrInvalidMasterKeyLength = fmt.Errorf("invalid key length, accepted key length is %d bytes", MasterKeyLength)
+
+// NewSearchable creates an Index which is used to manage keyword/ID pairs.
+func NewSearchable() Index {
+	mapping := make(map[string]SealedID)
+
+	return Index{mapping: mapping}
+}
+
+// Add is used to add a keyword/ID pair to the Index.
+func (i *Index) Add(key []byte, keyword, id string) error {
+	if len(key) != MasterKeyLength {
+		return ErrInvalidMasterKeyLength
+	}
+
+	k1 := crypto.KMACKDF(crypto.TaggerKeyLength, key, []byte("mac"), []byte(keyword))
+	k2 := crypto.KMACKDF(crypto.EncryptionKeyLength, key, []byte("id"), []byte(keyword))
+
+	tagger, err := crypto.NewKMAC256Tagger(k1)
+	if err != nil {
+		return err
+	}
+	cryptor, err := crypto.NewAESCryptor(k2)
+	if err != nil {
+		return err
+	}
+
+	count, err := i.count(key, keyword)
+	if err != nil {
+		return err
+	}
+	label, err := tagger.Tag(uint64Encode(count))
+	if err != nil {
+		return err
+	}
+
+	wrappedKey, encryptedID, err := cryptor.Encrypt([]byte(id), "")
```

Comment:
You should just be able to do `cryptor.Encrypt(&id, "")` and then similarly pass a `*string` to `Decrypt`.

Reply:
Fixed